Add custom animation frames to existing animations in Gamelab

caleybrock opened the pull request:
code.org spec: https://docs.google.com/document/d/1Yi3-dJ7-Y7gU2_mn_ZcunXR6lNQvXVpsZAGMKGzIxOo/edit#heading=h.rmbts45jim9j

<img width="1353" alt="screen shot 2017-01-10 at 3 52 00 pm" src="https://cloud.githubusercontent.com/assets/4640747/22390638/16d812d6-e4a1-11e6-8af7-f1d5ac7fe70b.png">
<img width="788" alt="screen shot 2017-01-10 at 3 51 15 pm" src="https://cloud.githubusercontent.com/assets/4640747/22390633/10eb0af4-e4a1-11e6-8bf1-6e6ddf8fc821.png">
<img width="765" alt="screen shot 2017-01-27 at 2 59 47 pm" src="https://cloud.githubusercontent.com/assets/4640747/22390670/4810e6c0-e4a1-11e6-811c-f299f9b389b8.png">

3 cases:
 - Add library animation as additional frames
 - Add blank frame
 - Add uploaded animations as additional frames

Relies on: https://github.com/code-dot-org/piskel/pull/24

caleybrock commented on apps/src/gamelab/animationListModule.js:
```diff
@@ -44,12 +44,57 @@ export const START_LOADING_FROM_SOURCE = 'AnimationList/START_LOADING_FROM_SOURC
+// pendingAnimationAddition is used for temporary storing additional
+// frames before they get added to the animation in Piskel.
+// pendingAnimationAddition gets added to animation in PiskelEditor.jsx
+function pendingAnimationAddition(state, action) {
```
I'm considering moving this out of AnimationList, but in some ways it makes a lot of sense here, because it will be an addition to an animation in the AnimationList. Thoughts?

caleybrock commented:
@islemaster here is the other one :)

Bjvanminnen commented on apps/src/gamelab/AnimationPicker/animationPickerModule.js:
```diff
@@ -1,6 +1,11 @@
+        addLibraryAnimation,
+        addBlankAnimationAddition,
+        addCustomAnimationAddition,
+        addLibraryAnimationAddition} from '../animationListModule';
```
Nit: should be indented two spaces (or maybe four?)

islemaster commented on apps/src/gamelab/AnimationPicker/animationPickerModule.js:
```diff
@@ -191,7 +196,7 @@ export function pickLibraryAnimation(animation) {
       dispatch(addLibraryAnimation(animation));
     } else if (goal === Goal.NEW_FRAME) {
-      // TODO (bbuchanan): Implement after integrating Piskel
+      dispatch(addLibraryAnimationAddition(animation));
```
:+1: :white_check_mark: Completing TODOs!

islemaster replied:
Suggested rename: `appendLibraryFrames`

Bjvanminnen commented on apps/src/gamelab/AnimationTab/PiskelEditor.jsx:
```diff
@@ -65,6 +70,39 @@ const PiskelEditor = React.createClass({
     if (newProps.selectedAnimation !== this.props.selectedAnimation) {
       this.loadSelectedAnimation_(newProps);
     }
+    if (newProps.pendingAnimationAddition
```
nit, i believe our style guide prefers the && to be on the preceding line, rather than starting a new one. also, it should be indented 4 spaces (to differentiate it from the code inside the block on line 75)

Bjvanminnen commented on apps/src/gamelab/AnimationTab/PiskelEditor.jsx:
```diff
@@ -65,6 +70,39 @@ const PiskelEditor = React.createClass({
       this.loadSelectedAnimation_(newProps);
     }
+    if (newProps.pendingAnimationAddition
+      && newProps.selectedAnimation === newProps.pendingAnimationAddition.key) {
```
Do we have to compare `newProps.pendingAnimationAddition` with `this.props.pendingAnimationAddition` or do we end up clearing this out after we've added it?

caleybrock replied:
we end up clearing it once it's done adding

Bjvanminnen commented on apps/src/gamelab/AnimationTab/PiskelEditor.jsx:
```diff
@@ -65,6 +70,39 @@ const PiskelEditor = React.createClass({
+        animationProps.dataURI,
+        animationProps.props.frameSize.x,
+        animationProps.props.frameSize.y,
+        animationProps.props.frameDelay,
```
`animationProps.props` Are we able to differentiate these different sets of props?

Bjvanminnen commented on apps/src/gamelab/animationListModule.js:
```diff
@@ -44,12 +44,56 @@ export const START_LOADING_FROM_SOURCE = 'AnimationList/START_LOADING_FROM_SOURC
+  pendingAnimationAddition
 });
 
+// pendingAnimationAddition is used for temporary storing additional
```
s/temporary/temporarily?

Bjvanminnen commented on apps/src/gamelab/animationListModule.js:
```diff
@@ -291,6 +335,20 @@ export function addBlankAnimation() {
 
+export function addBlankAnimationAddition() {
+  return (dispatch, getState) => {
+    var selectedAnimationKey = getState().animationTab.selectedAnimation;
```
nit: s/var/const

Bjvanminnen commented on apps/src/gamelab/animationListModule.js:
```diff
@@ -315,6 +373,19 @@ export function addAnimation(key, props) {
+ */
+export function addCustomAnimationAddition(props) {
+  return (dispatch, getState) => {
+    var selectedAnimationKey = getState().animationTab.selectedAnimation;
```
nit: s/var/const (here and elsewhere)

Bjvanminnen commented on apps/src/gamelab/animationListModule.js:
```diff
@@ -336,6 +407,46 @@ export function addLibraryAnimation(props) {
+
+/**
+ * After an animation addition is added, remove it as pending.
+ * @returns {{type: string}}
```
Not strictly true. We're returning a function, that eventually dispatches an object with type: string (here and elsewhere)

islemaster commented on apps/src/gamelab/AnimationPicker/animationPickerModule.js:
```diff
@@ -119,7 +124,7 @@ export function handleUploadComplete(result) {
         dispatch(addAnimation(key, animation));
       } else if (goal === Goal.NEW_FRAME) {
-        // TODO (bbuchanan): Implement after integrating Piskel
+        dispatch(addCustomAnimationAddition(animation));
```
Suggested rename: `appendCustomFrames`

islemaster commented on apps/src/gamelab/AnimationPicker/animationPickerModule.js:
```diff
@@ -172,7 +177,7 @@ export function pickNewAnimation() {
       dispatch(addBlankAnimation());
     } else if (goal === Goal.NEW_FRAME) {
-      // TODO (bbuchanan): Implement after integrating Piskel
+      dispatch(addBlankAnimationAddition());
```
Suggested rename: `appendBlankFrame`

Bjvanminnen commented on apps/src/gamelab/animationListModule.js:
```diff
@@ -291,6 +335,20 @@ export function addBlankAnimation() {
+      props: {
+        blankFrame: true
+      }
+    });
```
Here and elsewhere:

Not sure if Gamelab has been doing differently, but elsewhere we've tried to have individual action creators for each action, so we'd have something like:
```
dispatch(setPendingAnimationAddition(selectedAnimationKey))
```
and elsewhere
```
function (key) {
  return {
    type: SET_PENDING_ANIMATION,
    key,
    props: {
      blankFrame: true
    }
}
```

Bjvanminnen commented on apps/src/gamelab/animationListModule.js:
```diff
@@ -315,6 +373,19 @@ export function addAnimation(key, props) {
+    var selectedAnimationKey = getState().animationTab.selectedAnimation;
+    dispatch(setPendingAnimationAddition(selectedAnimationKey, props));
+    dispatch(loadAnimationAdditionFromSource(selectedAnimationKey, props));
+    dashboard.project.projectChanged();
```
I suspect we could directly import `projectChanged` from project.js instead of using `dashboard.project.projectChanged()`? (here and elsewhere)

caleybrock commented on apps/package.json:
```diff
@@ -39,7 +39,7 @@
     "@code-dot-org/bramble": "0.1.11",
     "@code-dot-org/p5.play": "1.1.3-cdo",
-    "@code-dot-org/piskel": "0.8.10",
+    "@code-dot-org/piskel": "0.9.1",
```
skipped a deprecated version that published without icons.

caleybrock commented:
Did a bunch of renaming, added action creator functions, wrote unit tests for the 3 types of new frames in `animationListModule`, used the new version of piskel, and fixed other style issues you suggested. @Bjvanminnen can you take another look?

Bjvanminnen commented on apps/src/gamelab/animationListModule.js:
```diff
@@ -44,12 +44,55 @@ export const START_LOADING_FROM_SOURCE = 'AnimationList/START_LOADING_FROM_SOURC
+    case START_LOADING_PENDING_FRAMES_FROM_SOURCE:
+      return Object.assign({}, state, {
+        loadedFromSource: false
+      });
```
Dont necessarily have to change this, but now that we have object destructuring, I think the preferred syntax here is:
```js
return {
  ...state,
  loadedFromSource: false
};
```

caleybrock replied:
This syntax is used throughout gamelab files, so I think I'll do a separate cleanup.

Bjvanminnen replied:
It's used throughout all of our code. Feel free to not cleanup existing usages if you don't feel like it, but going forward the new syntax is slightly preferred :)

Bjvanminnen commented on apps/src/gamelab/animationListModule.js:
```diff
@@ -300,17 +353,28 @@ export function addAnimation(key, props) {
-      key,
-      props
-    });
+    props.looping = true;
```
Our action creator probably shouldn't be mutating the input arguments. Might be better to instead just do:
`dispatch(addAnimationAction(key, { ...props, looping: true }))`.
Also, should the above two TODOs be addressed?

Bjvanminnen commented on apps/src/gamelab/animationListModule.js:
```diff
@@ -321,17 +385,38 @@ export function addAnimation(key, props) {
+  return dispatch => {
+    dispatch(removePendingFramesAction());
+  };
+}
```
This method is unnecessary. We can just export `removePendingFramesAction` and dispatch that directly from our component.

Bjvanminnen commented on apps/src/gamelab/animationListModule.js:
```diff
@@ -300,17 +353,28 @@ export function addAnimation(key, props) {
+ * @returns {function}
+ */
+export function appendCustomFrames(props) {
+  return (dispatch, getState) => {
```
nit: in cases where you don't actually use getState at all, i'd just make this `return dispatch => {`

caleybrock replied:
good to know - looks like this one uses getState though :)

Bjvanminnen replied:
haha, oops. i dont read so good.
